sensors/mag/lis2mdl.c

```diff
@@ -0,0 +1,229 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Driver for lis2mdl magnetometer
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/time.h>
+#include <sys/threads.h>
+#include <spi.h>
+#include <sensors-spi.h>
+#include <string.h>
+
+#include "../sensors.h"
+
+/* self-identification register of magnetometer */
+#define REG_WHOAMI 0x4f
+#define VAL_WHOAMI 0x40
+
+#define SPI_READ_BIT 0x80
+
+#define REG_CFG_REG_A               0x60
+#define VAL_CFG_REG_A_COMP_TEMP_EN  0x80
+#define VAL_CFG_REG_A_ODR_10HZ      0x00
+#define VAL_CFG_REG_A_ODR_20HZ      0x04
+#define VAL_CFG_REG_A_ODR_50HZ      0x08
+#define VAL_CFG_REG_A_ODR_100HZ     0x0c
+#define VAL_CFG_REG_A_MD_CONTINUOUS 0x00
+#define VAL_CFG_REG_A_MD_SINGLE     0x01
+#define VAL_CFG_REG_A_MD_IDLE       0x03
+
+#define REG_CFG_REG_B         0x61
+#define VAL_CFG_REG_B_LPF_OFF 0x00
+#define VAL_CFG_REG_B_LPF_ON  0x01
+
+#define REG_CFG_REG_C         0x62
+#define VAL_CFG_REG_C_4WSPI   0x04
+#define VAL_CFG_REG_C_I2C_DIS 0x20
+
+#define REG_DATA_OUT       0x68
+#define SENSOR_OUTPUT_SIZE 6
+
+#define MAG_CONV_MGAUSS 1.5F  /* conversion from sensor value to milligauss (equal to 10^-7 T) */
+#define MAG_OVERFLOW    21843 /* sensorhub will overflow if lis2mdl returns abs(raw) > MAG_OVERFLOW */
+
+
+typedef struct {
+	spimsg_ctx_t spiCtx;
+	oid_t spiSS;
+	sensor_event_t evt;
+	char stack[512] __attribute__((aligned(8)));
+} lis2mdl_ctx_t;
+
+
+static int16_t translateMag(uint8_t hbyte, uint8_t lbyte)
+{
+	int16_t val;
+
+	val = (hbyte << 8) | lbyte;
+
+	if (val > MAG_OVERFLOW) {
+		return MAG_OVERFLOW * MAG_CONV_MGAUSS;
+	}
+	else if (val < -MAG_OVERFLOW) {
+		return -MAG_OVERFLOW * MAG_CONV_MGAUSS;
+	}
+
+	/* conversion to milligaus (10^-7 T), as it is already sensor manager storage scale */
+	return val * MAG_CONV_MGAUSS;
```

Comment:
MISRA lesser evil fix
```suggestion
	else {
		/* conversion to milligaus (10^-7 T), as it is already sensor manager storage scale */
		return val * MAG_CONV_MGAUSS;
	}
```